### PR TITLE
Dynamic Subscription (BONUS: Allocators): rosidl_dynamic_typesupport

### PR DESCRIPTION
**Pull request by methylDragon**
See: https://github.com/ros2/ros2/issues/1405

### CODE REVIEW

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h [hunk 1]
```diff
@@ -48,36 +48,37 @@ extern "C"
...
  *
- * Ownership:
- *   - The `rosidl_message_type_support_t *` returned from this function has different ownership
- *     rules compared to the statically allocated `rosidl_message_type_support_t` structs from
- *     code-generated types!
- *    - The caller is responsible for deallocating the returned pointer
+ * The `type_hash`, `type_description`, `type_description_sources`, and `allocator` arguments are
+ * copied.
+
+ * The caller is responsible for deallocating the returned pointer.
```
Comment (wjwwood) [inline include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h]: What does this mean? Which pointer is returned?

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h [hunk 2]
```diff
@@ -89,21 +90,53 @@ typedef struct rosidl_dynamic_message_type_support_impl_s
...
+/// `rosidl_dynamic_message_type_support_handle_init()`, which has dynamically allocated members
 ///
 /// NOTE: Using this on a statically allocated typesupport will cause undefined behavior!
 ///       (Static memory will get freed in that case.)
 ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
 rcutils_ret_t
-rosidl_dynamic_message_type_support_handle_destroy(rosidl_message_type_support_t * ts);
+rosidl_dynamic_message_type_support_handle_fini(rosidl_message_type_support_t * ts);
+
+/// Initialized a `rosidl_dynamic_message_type_support_impl_t` with bound message description
```
Comment (wjwwood) [inline include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h]: What does `bound` mean in this context?

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h [hunk 3]
```diff
@@ -89,21 +90,53 @@ typedef struct rosidl_dynamic_message_type_support_impl_s
...
+/**
+ * The `type_hash`, `type_description`, `type_description_sources`, and `allocator` arguments are
+ * copied.
+ *
+ * The caller is responsible for deallocating the returned pointer.
```
Comment (wjwwood) [inline include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h]: Again, not sure what you're talking about here. Do you mean the `ts_impl`? If so, I think it's perhaps obvious that this is the case. They have to create something to pass in, so they have to clean it up. It might be that they pass in a reference to a thing created on the stack, in which case there's no deallocating. Do you instead mean that they need to call fini on it?

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h [hunk 4]
```diff
@@ -89,21 +90,53 @@ typedef struct rosidl_dynamic_message_type_support_impl_s
...
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ */
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_message_type_support_handle_impl_init(
```
Comment (wjwwood) [inline include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h:125]: Does this need to be public? Put another away does anyone ever call this other than the `rosidl_dynamic_message_type_support_handle_init()` function? If not, then maybe this could be private? Not sure, just asking.

**Review by wjwwood — APPROVED** [review]
A few comments, otherwise lgtm

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h (hunk 3 above)
Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h]: Right, the context has changed, I need to update the docstrings..

I can probably remove that line, since now the caller just has to call fini.

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h (hunk 2 above)
Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h]: I think I should rephrase it to `encapsulated`

File: include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h (hunk 4 above)
Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h:125]: Currently, no, but potentially yes (if someone wants to swap type support impls instead of remaking message type supports, or if we want to do partial instantiations (e.g. when we don't have descriptions yet))

Comment (wjwwood) [reply include/rosidl_dynamic_typesupport/dynamic_message_type_support_struct.h:125]: Ok, no problem, just double checking.

File: include/rosidl_dynamic_typesupport/api/dynamic_data.h [hunk 5]
```diff
@@ -109,21 +143,29 @@ rosidl_dynamic_typesupport_dynamic_data_get_name(
...
 rosidl_dynamic_typesupport_dynamic_data_clone(
   const rosidl_dynamic_typesupport_dynamic_data_t * other_dynamic_data,
-  rosidl_dynamic_typesupport_dynamic_data_t ** dynamic_data);  // OUT
+  rcutils_allocator_t * allocator,
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_fini(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data);
```
Comment (wjwwood) [inline include/rosidl_dynamic_typesupport/api/dynamic_data.h:168]: For other reviewers, the vast majority of changes for this pr can be summed up with this diff, basically adding allocators where we do allocation and avoiding double pointers as out parameters (matching how most of the `rcl`-like interfaces work, and the rest is mostly mechanical changes related to this shift.

**Review by wjwwood — APPROVED** [review]
lgtm, I started reviewing this last night and finished it this morning. Though there are a lot of lines changed, they are just repeating for the most part.

File: include/rosidl_dynamic_typesupport/api/dynamic_data.h [hunk 6]
```diff
@@ -22,12 +22,39 @@
...
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rosidl_dynamic_typesupport_dynamic_data_impl_t
+rosidl_dynamic_typesupport_get_zero_initialized_dynamic_data_impl(void);
+
+// Dynamic Data
+struct rosidl_dynamic_typesupport_dynamic_data_s
+{
+  rcutils_allocator_t allocator;
+  rosidl_dynamic_typesupport_dynamic_data_impl_t impl;
```
Comment (emersonknapp) [inline include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: Not using hidden-type `pimpl`?

File: src/api/dynamic_data.c [hunk 7]
```diff
@@ -130,35 +156,39 @@ rosidl_dynamic_typesupport_dynamic_data_get_array_index(
...
 rcutils_ret_t
 rosidl_dynamic_typesupport_dynamic_data_loan_value(
   rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
   rosidl_dynamic_typesupport_member_id_t id,
-  rosidl_dynamic_typesupport_dynamic_data_t ** loaned_dynamic_data)
+  rcutils_allocator_t * allocator,
+  rosidl_dynamic_typesupport_dynamic_data_t * loaned_dynamic_data)
 {
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(dynamic_data, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(allocator, RCUTILS_RET_INVALID_ARGUMENT);
```
Comment (emersonknapp) [inline src/api/dynamic_data.c:172]: I think you need to check `rcutils_allocator_is_valid(allocator)` as part of argument checking (for all) - allocator could be non-null but contain null function pointers.

File: include/rosidl_dynamic_typesupport/api/dynamic_data.h (hunk 6 above)
Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: I was thinking of doing that, but I thought it would be better to enforce that the impl is a complete type (and then defer it to the handle), so that we can enforce that the impl type has an allocator

Comment (clalancette) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: But isn't the the type defined just above?  That is, this implementation is completely in control of that structure, and thus can guarantee that it always has an allocator, no?

File: src/api/dynamic_data.c (hunk 7 above)
Comment (methylDragon) [reply src/api/dynamic_data.c:172]: I might defer this so we don't invalidate the windows debug CI run :x

(I fixed it)

File: src/api/dynamic_data.c [hunk 8]
```diff
@@ -302,28 +370,24 @@ rosidl_dynamic_typesupport_dynamic_data_destroy(
...
-  bool * success)
+  rcutils_uint8_array_t * buffer)
 {
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(dynamic_data, RCUTILS_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(buffer, RCUTILS_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(success, RCUTILS_RET_INVALID_ARGUMENT);
-  return (dynamic_data->serialization_support->methods->dynamic_data_serialize)(
-    dynamic_data->serialization_support->impl, dynamic_data->impl, buffer, success);
+  return (dynamic_data->serialization_support->methods.dynamic_data_serialize)(
+    &dynamic_data->serialization_support->impl, &dynamic_data->impl, buffer);
```
Comment (clalancette) [inline src/api/dynamic_data.c:394]: Don't we also need a pull request in https://github.com/ros2/rosidl_dynamic_typesupport_fastrtps to update `fastrtps__dynamic_data_serialize` as well?

(same goes for deserialize)

**Review by clalancette — CHANGES_REQUESTED** [review]
The general idea seems OK to me, but I think we are missing a PR here (or I'm misunderstanding something).

File: include/rosidl_dynamic_typesupport/api/dynamic_data.h (hunk 6 above)
Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: ah, right.. I think I'm avoiding the use of a pointer so we don't have to keep doing null checks on the impl member as well..

The impl is unusable without the associated serialization support in either case, so in my view composition makes sense

File: src/api/dynamic_data.c (hunk 8 above)
Comment (methylDragon) [reply src/api/dynamic_data.c:394]: You're right, I forgot to spin up the PR.. I thought I caught them all :man_facepalming: 
Sorry about that

https://github.com/ros2/rosidl_dynamic_typesupport_fastrtps/pull/3

File: include/rosidl_dynamic_typesupport/api/dynamic_data.h (hunk 6 above)
Comment (clalancette) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: Yeah, actually, I largely agree with you.  The only problem here is that if we want to add, remove, or change structure members in the implementation, we won't be able to backport it due to ABI concerns.  But if we are OK with that limitation for now, we can leave it as-is.

Comment (methylDragon) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: The implementation contains a void * handle, so that's a good escape hatch I think?

Comment (clalancette) [reply include/rosidl_dynamic_typesupport/api/dynamic_data.h:50]: Yeah, it's good enough for now.  It isn't our usual style, but it will do in a pinch.

**Review by clalancette — APPROVED** [review]
